common/notification/messages.ts

```diff
@@ -42,6 +48,68 @@ export async function getMessageForNotification(
     return { body, headline, modalText, navigateTo: translation.navigateTo, type: notification.type as NotificationType };
 }
 
+type CachedMessageTemplate = {
+    type: NotificationType;
+    body: (context: NotificationContext) => string;
+    headline: (context: NotificationContext) => string;
+    navigateTo?: (context: NotificationContext) => string;
+    modalText?: (context: NotificationContext) => string;
+};
+
+const cachedTemplates = new Map</* notification id + language */ string, CachedMessageTemplate | null /* = no message */>();
+
+async function loadTemplate(notificationID: number, language: TranslationLanguage): Promise<CachedMessageTemplate | null> {
+    const cacheKey = notificationID + ':' + language;
+
+    if (cachedTemplates.has(cacheKey)) {
+        logger.debug(`Got Message for Notification(${notificationID}) from cache`);
+        return cachedTemplates.get(cacheKey);
+    }
+
+    const message = await getMessageForNotification(notificationID, language);
+
+    if (!message) {
+        cachedTemplates.set(cacheKey, null);
+        return null;
+    }
+
+    const { type, headline, body, modalText, navigateTo } = message;
+
+    const template = {
+        type,
+        body: compileTemplate(body),
+        headline: compileTemplate(headline),
+        navigateTo: navigateTo && compileTemplate(navigateTo),
+        modalText: modalText && compileTemplate(modalText),
+    };
+
+    cachedTemplates.set(cacheKey, template);
+    logger.info(`Loaded Message for Notification(${notificationID}) into cache`);
+    return template;
+}
+
+// Renders a Message for a certain Concrete Notification
+export async function getMessage(
+    concreteNotification: ConcreteNotification,
+    language: TranslationLanguage = TranslationLanguage.DE
+): Promise<NotificationMessageType | null> {
+    const template = await loadTemplate(concreteNotification.notificationID, language);
+    if (!template) {
+        return null;
+    }
+
+    const legacyUser = await getUserTypeORM(concreteNotification.userId);
```

Comment:
/nit Should we use prisma to push the migration forward?

Reply:
yes, but touching the noitification system in that way might potentially break all notifications, so I want to do that refactoring in a separate step.